Update gateways info

ElPaisano opened the pull request:
Incorporates some of the content from https://blog.ipfs.tech/2022-06-30-practical-explainer-ipfs-gateways-2/ into existing GW concepts and ref material 

Since the blog + the existing https://docs.ipfs.tech/concepts/ipfs-gateway/ page both have alot of contents, this PR:
- Prunes out best practice and troubleshooting type info from concepts/ipfs-gateway/
- Puts some best practices in /how-to/gateway-best-practices/
- Puts troubleshooting info in /how-to/gateway-troubleshooting/

2color commented on docs/how-to/gateway-best-practices.md:
```diff
@@ -0,0 +1,123 @@
+- Monitor disk I/O and make sure that no other processes are causing disk I/O bottlenecks with a tool like [iotop](https://linux.die.net/man/1/iotop) or [iostat](https://linux.die.net/man/1/iostat).
+
+
+Using this remote pinning API, you can implement pinning to multiple services (opens new window)as part of your workflow for uploading immutable data to IPFS.
```
```suggestion
```

Remove this line as its a duplicate of the one above.

2color commented on docs/concepts/ipfs-gateway.md:
```diff
@@ -50,23 +65,19 @@ A gateway behind a firewall represents just one potential location for a private
 
 - [Read/write support](#read-only-and-writeable-gateways)
 - [Authentication support](#authenticated-gateways)
 - [Resolution style](#resolution-style)
```
In this introduction, I'd add a red warning for **path resolution** because it does not provide origin isolation.

2color commented on docs/how-to/gateway-best-practices.md:
```diff
@@ -0,0 +1,123 @@
+
+Trusting a specific gateway, in turn, requires you to trust the gateway's issuing Certificate Authorities and the security of the public key infrastructure employed by that gateway. Compromised certificate authorities or public-key infrastructure implementations may undermine the trustworthiness of the gateway.
+
+## Violation of same-origin policy
```
```suggestion
## Use subdomain gateway resolution for origin isolation
```

2color commented on docs/how-to/gateway-best-practices.md:
```diff
@@ -0,0 +1,123 @@
+
+To prevent one website from improperly accessing HTTP session data associated with a different website, the [same-origin policy](https://en.wikipedia.org/wiki/Same-origin_policy) permits script access only to pages that share a common domain name and port.
+
+Consider two web pages stored in IPFS: `ipfs://{CID A}/{webpage A}` and `ipfs://{CID B}/{webpage B}`. Code on `webpage A` should not access data from `webpage B`, as they do not share the same content ID (origin).
```
```suggestion
Consider two CIDs each representing a different website accessed with the path resolution style: 
 - `https://ipfs.io/{CID A}/{website A}`
 - `https://ipfs.io/{CID B}/{website B}`. 
 
Because their origin (hostname and port) are the same, the same-origin policy does not apply. 

To ensure the security provided by the same-origin policy, use the subdomain gateway:
```bash
https://{CID A}.ipfs.{gatewayURL}/{website A}
https://{CID B}.ipfs.{gatewayURL}/{website B}
```

```

2color commented on docs/how-to/gateway-troubleshooting.md:
```diff
@@ -0,0 +1,144 @@
+
+# Troubleshooting HTTP Gateways
+
+IPFS HTTP Gateways (GWs) provide an HTTP-based service that allows IPFS-incompatible browsers, tools and software to access IPFS content. When using HTTP GWs , developers may need to troubleshoot issues like a:
```
```suggestion
IPFS HTTP Gateways are an HTTP-based service allowing browsers, tools and software to retrieve IPFS content with HTTP. When using HTTP Gateways, developers may need to troubleshoot issues like a:
```

Let's stick to Gateways. I don't think we need to shorten it.

2color reviewed (commented):
Thanks for bringing those into the docs. I left some comments on changes since the blog was published.

We should more strongly recommend using subdomain gateway resolution as it provides more security.

2color commented on docs/how-to/gateway-best-practices.md:
```diff
@@ -0,0 +1,123 @@
+
+Various best practices for the use of IPFS gateways are listed below. To learn more about the concepts behind IPFS gateways, including how they work, available providers, types and FAQs, see [IPFS Gateway](../concepts/ipfs-gateway.md). For troubleshooting information, see [Troubleshooting](./gateway-troubleshooting.md).
+
+## When not to use a gateway
```
I would move this whole section to the bottom of the page to avoid the best practices starting with when not to use a gateway.

I think this "Delay-sensitive applications" section is questionable. Given that CIDs are immutable, content on IPFS plays really well with HTTP caching. 

If you're publishing a website with IPFS, using a gateway with subdomain resolution works pretty darn well (and fast) if you're loading from the IPFS node providing the content. 

Embedding IPFS into your app doesn't really work for the web/browser-based stuff — for that, we need deeper browser support.

azimanuar5874 commented:
![Y block the ME r G](https://user-images.githubusercontent.com/73795636/227302268-cf9fe8b9-7b2d-45c6-9f50-ac909808ba31.png)

High all. My name is Double A. And im here just to make this world save to us all to live. And for sure, i dont want to mess with ai pi as F with the A.P.I anymore. Im here not to read all the COMPLEXcity that u all just made! thats all. Make your own choice, O.K :)))

azimanuar5874 commented on docs/how-to/gateway-best-practices.md:
![howdomainswork](https://user-images.githubusercontent.com/73795636/227304051-78aae14c-667e-4d83-8581-649a203f1998.png)

which DOMain that ACTu(ally) works with all this kompleks things. And lets make it clear when we talk about the future of sick V or web TR33 or VVVVVV .

ElPaisano commented:
> Thanks for bringing those into the docs. I left some comments on changes since the blog was published.
> 
> We should more strongly recommend using subdomain gateway resolution as it provides more security.

@2color I think I addressed your comments in the updates

hacdias reviewed (commented):
I didn't have time yet to go over the whole document, but I noticed there aren't any mentions for the gateway specifications. I would add some mention that the specification for the different types and parts of the gateways are available at https://specs.ipfs.tech/http-gateways/ - that'd be great visibility with information for those who want to implement their own gateway.

hacdias commented on docs/concepts/ipfs-gateway.md:
```diff
@@ -11,28 +11,43 @@ related:
+- **If the CID is cached locally**, the gateway responds with the content referred to by the CID, and the lifecycle is complete.
 
-IPFS deployment seeks to include native support of IPFS in all popular browsers and tools. Gateways provide workarounds for applications that do not yet support IPFS natively. For example, errors occur when a browser that does not support IPFS attempts access to IPFS content in the canonical form of `ipfs://{CID}/{optional path to resource}`. Other tools that rely solely on HTTP encounter similar ...
+- **If the CID is not in the local cache**, the gateway will attempt to retrieve it from the network.
```
> the gateway will attempt to retrieve it from the network

That is the default of Kubo. However, there might be gateways that only serve content that they have and/or want to provide. For example, a Kubo gateway with `NoFetch` enabled will **not** attempt to retrieve content from the network. I'm not sure if we should keep this, or remove it. Opinions?

hacdias commented on docs/concepts/ipfs-gateway.md:
```diff
@@ -50,26 +65,26 @@ A gateway behind a firewall represents just one potential location for a private
 
 - Protocol Labs, which deploys the public gateway `https://ipfs.io`.
-- Third-party public gateways. E.g., `https://cf-ipfs.com`.
+- Third-party public gateways such as `https://cf-ipfs.com`.
```
```suggestion
- Third-party public gateways, such as `https://cf-ipfs.com`.
```

ElPaisano commented:
> I didn't have time yet to go over the whole document, but I noticed there aren't any mentions for the gateway specifications. I would add some mention that the specification for the different types and parts of the gateways are available at https://specs.ipfs.tech/http-gateways/ - that'd be great visibility with information for those who want to implement their own gateway.

Thanks for reviewing @hacdias , much appreciated

I added the spec link in a few place that made sense. RE your first comment https://github.com/ipfs/ipfs-docs/pull/1531#discussion_r1155579068, I added a note at the beginning of the lifecycle section calling out your point, so readers are aware of these nuances. See https://github.com/ipfs/ipfs-docs/pull/1531/commits/bb394f26baa17ea21dd66917004b29d0ea7a69b8. 

If we need to refine this section, we can do it in future PRs, as this one has been sitting for awhile now

2color commented on docs/how-to/gateway-best-practices.md:
```diff
@@ -0,0 +1,112 @@
+
+
+
+## Avoiding centralization
```
This paragraph is too convoluted.

The tile could be much more specific and direct: **avoid hardcoding a gateway URL, e.g. ipfs.io/ipfs/CID** in places where possible! It must also acknowledge that it's not feasible in some situations, e.g. embedding image from IPFS on web and that can be worked around by owning the domain and proxying.

2color commented on docs/how-to/gateway-best-practices.md:
```diff
@@ -0,0 +1,112 @@
+
+Trusting a specific gateway, in turn, requires you to trust the gateway's issuing Certificate Authorities and the security of the public key infrastructure employed by that gateway. Compromised certificate authorities or public-key infrastructure implementations may undermine the trustworthiness of the gateway.
+
+## Use subdomain gateway resolution for origin isolation
```
Should be the first one.

2color reviewed (approved):
This is a great improvement. Wonderful to see the new specs page https://specs.ipfs.tech/http-gateways/  linked!

Besides some comments about the best practices, this is good to go.

![](https://media.giphy.com/media/gIpfiHhGbEkLYqozsZ/giphy.gif)
